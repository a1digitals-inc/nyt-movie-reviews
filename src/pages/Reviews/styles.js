import styled from 'styled-components';
import { darken } from 'polished';

export const Header = styled.div`
  background: #e6e6e6;
  padding: 0 30px;
`;

export const Filter = styled.div`
  max-width: 900px;
  display: flex;
  align-self: center;
  align-items: center;
  margin: 0 auto;

  form {
    width: 100%;
    display: flex;
    flex-flow: row wrap;
    flex-direction: column;
    align-items: stretch;
    justify-content: space-between;

    input,
    select {
      background: rgba(0, 0, 0, 0.1);
      border: 0;
      border-radius: 4px;
      padding: 10px 15px;
      margin: 15px 0 15px 0;
    }

    label {
      input,
      span {
        font-size: smaller;
        margin: 0 5px 0 0;
        padding: 0;
      }

      input {
        vertical-align: sub;
      }
    }

    button {
      height: 44px;
      margin: 15px 0 15px 0;
      padding: 10px 20px;
      border: 0;
      border-radius: 4px;
      font-size: 16px;
      font-weight: bold;
      color: #fff;
      background: #999;
      transition: background 0.2s;

      &:hover {
        background: ${darken(0.5, '#999')};
      }
    }

    @media (min-width: 900px) {
      flex-direction: row;
      align-items: center;
    }
  }
`;

export const Container = styled.div`
  max-width: 1020px;
  display: flex;
  flex-direction: column;
  margin: 0 auto;
  padding: 0 20px 50px;
`;
