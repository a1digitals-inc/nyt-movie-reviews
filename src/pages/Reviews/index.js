import React, { useState, useEffect, useCallback } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useParams } from 'react-router-dom';

import PageLoading from '../../components/PageLoading';
import NotFound from '../../components/NotFound';
import Review from '../../components/Review';
import ComponentLoading from '../../components/ComponentLoading';
import useInfiniteScroll from '../../hooks/useInfiniteScroll';
import {
  searchReviewsRequest,
  searchMoreReviewsRequest,
} from '../../store/modules/reviews/actions';
import { Header, Filter, Container } from './styles';

const Reviews = () => {
  const { critic } = useParams();

  const criticParam = critic || '';

  const dispatch = useDispatch();

  const initialLoading = useSelector(state => state.reviews.initialLoading);
  const loadingMore = useSelector(state => state.reviews.loadingMore);
  const filters = useSelector(state => state.reviews.filters);
  const reviews = useSelector(state => state.reviews.reviews);
  const hasMore = useSelector(state => state.reviews.hasMore);
  const hasError = useSelector(state => state.reviews.hasError);

  const [componentFilters, setComponentFilters] = useState(filters);

  useEffect(() => {
    const initialFilters = {
      order: '',
      query: '',
      reviewer: criticParam,
      criticsPick: '',
    };

    setComponentFilters(initialFilters);

    dispatch(searchReviewsRequest(initialFilters));
  }, [dispatch, criticParam]);

  const searchMoreReviewsCallback = useCallback(() => {
    if (hasMore && !hasError) {
      dispatch(searchMoreReviewsRequest());
    }
  }, [dispatch, hasMore, hasError]);

  useInfiniteScroll(searchMoreReviewsCallback);

  function handleFilterChange(e) {
    e.preventDefault();

    const {
      target: { name, checked, value },
    } = e;

    if (['criticsPick'].includes(name)) {
      setComponentFilters({
        ...componentFilters,
        [name]: checked,
      });
    } else {
      setComponentFilters({
        ...componentFilters,
        [name]: value,
      });
    }
  }

  async function handleFilterSubmit(e) {
    e.preventDefault();

    dispatch(searchReviewsRequest(componentFilters));
  }

  if (initialLoading) {
    return <PageLoading />;
  }

  return (
    <>
      <Header>
        <Filter>
          <form autoComplete="off" onSubmit={handleFilterSubmit}>
            <select
              name="order"
              onChange={handleFilterChange}
              value={componentFilters.order}
            >
              <option value="">Order</option>
              <option value="by-title">Order by Title</option>
              <option value="by-publication-date">
                Order by Publication date
              </option>
              <option value="by-opening-date">Order by Opening date</option>
            </select>
            <input
              type="text"
              name="query"
              placeholder="Keyword"
              onChange={handleFilterChange}
              value={componentFilters.query}
            />
            <input
              type="text"
              name="reviewer"
              placeholder="Reviewer"
              onChange={handleFilterChange}
              value={componentFilters.reviewer}
            />
            <label htmlFor="criticsPick">
              <input
                type="checkbox"
                id="criticsPick"
                name="criticsPick"
                onChange={handleFilterChange}
                checked={componentFilters.criticsPick}
              />
              <span>NYT Criticâ€™s Pick</span>
            </label>
            <button type="submit">Search</button>
          </form>
        </Filter>
      </Header>
      {!reviews.length ? (
        <NotFound Message="No records found" />
      ) : (
        <Container>
          {reviews.map(review => {
            return (
              <Review
                key={`${review.display_title}-${review.byline}`}
                rawData={review}
              />
            );
          })}
          {loadingMore && <ComponentLoading />}
        </Container>
      )}
    </>
  );
};

export default Reviews;
