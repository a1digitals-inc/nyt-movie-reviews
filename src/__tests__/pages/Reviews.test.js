import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { MemoryRouter } from 'react-router-dom';

import Reviews from '../../pages/Reviews';

describe('Reviews page', () => {
  const middlewares = [];
  const mockStore = configureStore(middlewares);
  const initialState = {
    reviews: {
      initialLoading: false,
      loadingMore: false,
      searchCounter: 0,
      filters: {
        order: '',
        query: '',
        reviewer: '',
        criticsPick: '',
      },
      reviews: [],
      hasMore: false,
      hasError: false,
    },
  };
  const store = mockStore(initialState);

  it('should render without crashing', () => {
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: '/reviews', key: 'testKey' }]}
        >
          <Reviews />
        </MemoryRouter>
      </Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });
});
