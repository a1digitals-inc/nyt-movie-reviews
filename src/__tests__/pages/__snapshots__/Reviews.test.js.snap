// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reviews page should render without crashing 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <MemoryRouter
    initialEntries={
      Array [
        Object {
          "key": "testKey",
          "pathname": "/reviews",
        },
      ]
    }
  >
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "canGo": [Function],
          "createHref": [Function],
          "entries": Array [
            Object {
              "hash": "",
              "key": "testKey",
              "pathname": "/reviews",
              "search": "",
            },
          ],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "index": 0,
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "key": "testKey",
            "pathname": "/reviews",
            "search": "",
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <Reviews>
        <styled.div>
          <StyledComponent
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "sc-htoDjs",
                  "isStatic": false,
                  "lastClassName": "ieqtDf",
                  "rules": Array [
                    "
  background: #e6e6e6;
  padding: 0 30px;
",
                  ],
                },
                "displayName": "styled.div",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "sc-htoDjs",
                "target": "div",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
          >
            <div
              className="sc-htoDjs ieqtDf"
            >
              <styled.div>
                <StyledComponent
                  forwardedComponent={
                    Object {
                      "$$typeof": Symbol(react.forward_ref),
                      "attrs": Array [],
                      "componentStyle": ComponentStyle {
                        "componentId": "sc-dnqmqq",
                        "isStatic": false,
                        "lastClassName": "cDUlEM",
                        "rules": Array [
                          "
  max-width: 900px;
  display: flex;
  align-self: center;
  align-items: center;
  margin: 0 auto;

  form {
    width: 100%;
    display: flex;
    flex-flow: row wrap;
    flex-direction: column;
    align-items: stretch;
    justify-content: space-between;

    input,
    select {
      background: rgba(0, 0, 0, 0.1);
      border: 0;
      border-radius: 4px;
      padding: 10px 15px;
      margin: 15px 0 15px 0;
    }

    label {
      input,
      span {
        font-size: smaller;
        margin: 0 5px 0 0;
        padding: 0;
      }

      input {
        vertical-align: sub;
      }
    }

    button {
      height: 44px;
      margin: 15px 0 15px 0;
      padding: 10px 20px;
      border: 0;
      border-radius: 4px;
      font-size: 16px;
      font-weight: bold;
      color: #fff;
      background: #999;
      transition: background 0.2s;

      &:hover {
        background: ",
                          "#191919",
                          ";
      }
    }

    @media (min-width: 900px) {
      flex-direction: row;
      align-items: center;
    }
  }
",
                        ],
                      },
                      "displayName": "styled.div",
                      "foldedComponentIds": Array [],
                      "render": [Function],
                      "styledComponentId": "sc-dnqmqq",
                      "target": "div",
                      "toString": [Function],
                      "warnTooManyClasses": [Function],
                      "withComponent": [Function],
                    }
                  }
                  forwardedRef={null}
                >
                  <div
                    className="sc-dnqmqq cDUlEM"
                  >
                    <form
                      autoComplete="off"
                      onSubmit={[Function]}
                    >
                      <select
                        name="order"
                        onChange={[Function]}
                        value=""
                      >
                        <option
                          value=""
                        >
                          Order
                        </option>
                        <option
                          value="by-title"
                        >
                          Order by Title
                        </option>
                        <option
                          value="by-publication-date"
                        >
                          Order by Publication date
                        </option>
                        <option
                          value="by-opening-date"
                        >
                          Order by Opening date
                        </option>
                      </select>
                      <input
                        name="query"
                        onChange={[Function]}
                        placeholder="Keyword"
                        type="text"
                        value=""
                      />
                      <input
                        name="reviewer"
                        onChange={[Function]}
                        placeholder="Reviewer"
                        type="text"
                        value=""
                      />
                      <label
                        htmlFor="criticsPick"
                      >
                        <input
                          checked=""
                          id="criticsPick"
                          name="criticsPick"
                          onChange={[Function]}
                          type="checkbox"
                        />
                        <span>
                          NYT Criticâ€™s Pick
                        </span>
                      </label>
                      <button
                        type="submit"
                      >
                        Search
                      </button>
                    </form>
                  </div>
                </StyledComponent>
              </styled.div>
            </div>
          </StyledComponent>
        </styled.div>
        <NotFound
          Code={null}
          Message="No records found"
        >
          <styled.div>
            <StyledComponent
              forwardedComponent={
                Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "attrs": Array [],
                  "componentStyle": ComponentStyle {
                    "componentId": "sc-ifAKCX",
                    "isStatic": false,
                    "lastClassName": "ioktai",
                    "rules": Array [
                      "
  display: flex;
  justify-content: center;
  align-items: center;
  font-style: oblique;
",
                    ],
                  },
                  "displayName": "styled.div",
                  "foldedComponentIds": Array [],
                  "render": [Function],
                  "styledComponentId": "sc-ifAKCX",
                  "target": "div",
                  "toString": [Function],
                  "warnTooManyClasses": [Function],
                  "withComponent": [Function],
                }
              }
              forwardedRef={null}
            >
              <div
                className="sc-ifAKCX ioktai"
              >
                <styled.div>
                  <StyledComponent
                    forwardedComponent={
                      Object {
                        "$$typeof": Symbol(react.forward_ref),
                        "attrs": Array [],
                        "componentStyle": ComponentStyle {
                          "componentId": "sc-EHOje",
                          "isStatic": false,
                          "lastClassName": "cTEuPM",
                          "rules": Array [
                            "
  display: flex;
  flex-direction: column;
  text-align: center;
  margin-top: 90px;
",
                          ],
                        },
                        "displayName": "styled.div",
                        "foldedComponentIds": Array [],
                        "render": [Function],
                        "styledComponentId": "sc-EHOje",
                        "target": "div",
                        "toString": [Function],
                        "warnTooManyClasses": [Function],
                        "withComponent": [Function],
                      }
                    }
                    forwardedRef={null}
                  >
                    <div
                      className="sc-EHOje cTEuPM"
                    >
                      <strong />
                      <strong>
                        No records found
                      </strong>
                    </div>
                  </StyledComponent>
                </styled.div>
              </div>
            </StyledComponent>
          </styled.div>
        </NotFound>
      </Reviews>
    </Router>
  </MemoryRouter>
</Provider>
`;
